crates/transpiler/src/passes/basis_translator/compose_transforms.rs

```diff
@@ -24,129 +36,174 @@ use qiskit_circuit::{
-                                .collect::<SmallVec<[Param; 3]>>(),
-                        )
-                    })
-                    .collect::<Vec<_>>();
-                for (node, params) in nodes_to_replace {
-                    let param_mapping: IndexMap<ParameterUuid, Param, ahash::RandomState> =
-                        equiv_params
+    for ((gate_name, gate_num_qubits), (equiv_params, equiv)) in basis_transforms {
+        for (_, dag) in &mut mapped_instructions.values_mut() {
+            let nodes_to_replace = dag
+                .op_nodes(true)
+                .filter(|(_, op)| {
+                    (op.op.num_qubits() == *gate_num_qubits) && (op.op.name() == gate_name.as_str())
+                })
+                .map(|(node, op)| {
+                    (
+                        node,
+                        op.params_view()
                             .iter()
-                            .map(|x| {
-                                // extract the Symbol from the Param
-                                let symbol = match x {
-                                    Param::ParameterExpression(expr) => {
-                                        expr.try_to_symbol().unwrap()
-                                    }
-                                    _ => panic!("Equivalence transform param must be expression."),
-                                };
+                            .cloned()
+                            .collect::<SmallVec<[Param; 3]>>(),
+                    )
+                })
+                .collect::<Vec<_>>();
+            for (node, params) in nodes_to_replace {
+                let param_mapping: IndexMap<ParameterUuid, Param, ahash::RandomState> =
+                    equiv_params
+                        .iter()
+                        .map(|x| match x {
+                            Param::ParameterExpression(parameter_expression) => {
+                                let symbol = parameter_expression.try_to_symbol().unwrap();
                                 ParameterUuid::from_symbol(&symbol)
-                            })
-                            .zip(params)
-                            .map(|(uuid, param)| (uuid, param.clone_ref(py)))
-                            .collect();
-                    let mut replacement = equiv.clone();
-                    replacement
-                        .0
-                        .assign_parameters_from_mapping(param_mapping)?;
-                    let replace_dag: DAGCircuit = DAGCircuit::from_circuit_data(
-                        &replacement.0,
-                        true,
-                        None,
-                        None,
-                        None,
-                        None,
-                    )?;
-                    dag.substitute_node_with_dag(node, &replace_dag, None, None, None)?;
-                }
+                            }
+                            _ => panic!("A non parameter-expression has suck in"),
```

Comment:
This should be unreachable, right? 😄 
```suggestion
                            _ => unreachable!("A non parameter-expression has snuck in"),
```

Reply:
Added in cbe4e1d868234c50cce87a47723cbac688aea0e7